### PR TITLE
Browse filter results on page

### PR DESCRIPTION
**Pull request by litvinovg**
[VIVO PR](https://github.com/vivo-project/VIVO/pull/4005)
# What does this pull request do?

- new page content type "Browse search filter results" to display search results limited by selected filter. Also additional filters and sort options could be defined in custom page template (template [example](https://github.com/TIBHannover/tib-fis-installer/blob/1.15.1-snapshot/project-installer/webapp/src/main/webapp/templates/freemarker/body/menupage/browsePersonSearchFilterValues.ftl))
- new facet autocomplete input field that appears if number of facets in bigger than more limit defined for the filter
- new property to limit variables of document modifiers to be populated into search indexes

# What's new?
Refactored PagedSearchController. 
Added search filter option to ManagePageGenerator
Refactored ProcessN3GetterDataMap
Created new data getter class display:SearchFilterValuesDataGetter , java implementation in SearchFilterValuesDataGetter.java
Created object property search:direction to configure sort direction of filter values and search results.
Created filters to filter result in alphabetical indexes 
:filter_label_regex - default regex filter that uses i18n field :field_label_display with name locale + _label_display
:filter_raw_label_regex - fallback regex filter that uses nameRaw search field for VIVO instances without i18n support
search:limitDisplayTo object property was created to limit user access to display of search groups, filters, values.
Resolved bug on search page which appears if user select too many filters. In that case search results are empty and selected filter values are not shown on the page, so the user can't unselect them. 
Added options to sort results, number of results on page and "Download results" options on custom pages

Applied changes for search ontology (from [ontology interests group meeting](https://wiki.lyrasis.org/display/VIVO/2025-01-08+Ontology+Interest+Group+Call))
Removed search:isAscending boolean data property,  new object property search:direction should be used instead. 
(search:isAscending is still could be used for backwards compatibility).
Removed search:reverseFacetOrder boolean data property, new object property search:direction should be used instead. 
(search:reverseFacetOrder is still could be used for backwards compatibility).
Renamed search:order integer data property into search:rank, 
(search:order is still could be used for backwards compatibility).
Created class search:SortDirection and two individuals search-ind:ascending and search-ind:descending to be used with search:direction object property.
Added facet autocomplete input field that is being visible in case filter has number of values greater than limit of facets to be shown before "more..." button

# How should this be tested?
How to test new page content type:
* Create a new page in Page management menu. 
* For the content select "Browse search filter results", select a type and save it.
* To select custom template select Template option "Custom template requiring content" and enter filename of custom Freemarker template in input field (Optional).
* Open page, try selecting facets, alphabetical index, pagination
* Test above in VIVO with and without language filtering

How to test facet autocomplete input
* To test facet autocomplete input you would need a filter that doesn't use URIs to store in search index. [An example](https://github.com/litvinovg/VIVO/compare/browse-filter-results...litvinovg:VIVO:facet-autocomplete-test-settings) of modification of organization filter
* Run full search re-index after filter modification.
* You can check how it work on /search page by selecting the filter and entering at least 3 characters in input field. 
* Suggestions should appear. Select a suggestion and page should reload with applied selected facet.
 
# Additional Notes:
Examples of customized [templates](https://github.com/TIBHannover/tib-fis-installer/tree/1.15.1-snapshot/project-installer/webapp/src/main/webapp/templates/freemarker/body/menupage), [document modifiers](https://github.com/TIBHannover/tib-fis-installer/blob/1.15.1-snapshot/project-installer/home/src/main/resources/rdf/display/everytime/searchFieldsConfigurationVivo.n3)
This change require documentation to be updated.

# Interested parties
@hauschke @VIVO-project/vivo-committers

# Reviewers' expertise

Candidates for reviewing this PR should have some of the following expertises:
1. Java
2. JavaScript
3. FreeMarker
4. SPARQL
5. Ontologies
1. Natural language knowledge
    1. English
    2. German
    3. Spanish
    4. French
    5. Portuguese
    6. Russian
    7. Serbian

# Reviewers' report template
**Please update the following template which should be used by reviewers.**
## General comment
A reviewer should provide here comments and suggestions for requested changes if any.
## Testing
A reviewer should briefly describe here how it was tested
## Code reviewing
A reviewer should briefly describe here which part was code reviewed

### CODE REVIEW

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/utils/dataGetter/SearchFilterValuesDataGetter.java [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+
+public class SearchFilterValuesDataGetter extends DataGetterBase implements DataGetter {
+    private static final String defaultTemplate = "browseSearchFilterValues.ftl";
+    private static final String searchFilterUri = "<" + DisplayVocabulary.SEARCH_FILTER_VALUE + ">";
+    private static final Log log = LogFactory.getLog(SearchFilterValuesDataGetter.class);
+
+    String dataGetterURI;
+    String searchFilter;
+    VitroRequest vreq;
+    ServletContext context;
```
Comment (chenejac) [inline api/src/main/java/edu/cornell/mannlib/vitro/webapp/utils/dataGetter/SearchFilterValuesDataGetter.java:44]: I know this is copied from other DataGetters, but I am wondering why without access modifier (private)? Is this really planned to be used by other classes from this package? I don't think so

File: home/src/main/resources/rdf/i18n/de_DE/interface-i18n/firsttime/vitro_UiLabel.ttl [hunk 2]
```diff
@@ -6445,3 +6445,28 @@ uil-data:password_reset_admin_notification_email_html.Vitro
         uil:hasApp       "Vitro" ;
         uil:hasKey       "password_reset_admin_notification_email_html" ;
         uil:hasPackage   "Vitro-languages" .
+
+uil-data:browse_results_alphabetical_index.Vitro
+        rdf:type         owl:NamedIndividual ;
+        rdf:type         uil:UILabel ;
+        rdfs:label       "a,b,c,d,e,f,g,h,i,j,k,l,m,n,o,p,q,r,s,t,u,v,w,x,y,z"@de-DE ;
```
Comment (chenejac) [inline home/src/main/resources/rdf/i18n/de_DE/interface-i18n/firsttime/vitro_UiLabel.ttl:6452]: should be extended with German letters. This might be a task for a native German speaker, but it looks to me this might be ä, ö, ü, ß at the end after z - [link](https://www.reddit.com/r/NoStupidQuestions/comments/z7ye4a/how_does_the_german_alphabetic_order_work/#:~:text=So%2C%20the%20alphabet%20goes%20from,not%20one%20after%20the%20other.)

File: home/src/main/resources/rdf/i18n/pt_BR/interface-i18n/firsttime/vitro_UiLabel.ttl [hunk 3]
```diff
@@ -6445,3 +6445,28 @@ uil-data:password_reset_admin_notification_email_html.Vitro
         uil:hasApp       "Vitro" ;
         uil:hasKey       "password_reset_admin_notification_email_html" ;
         uil:hasPackage   "Vitro-languages" .
+
+uil-data:browse_results_alphabetical_index.Vitro
+        rdf:type         owl:NamedIndividual ;
+        rdf:type         uil:UILabel ;
+        rdfs:label       "a,b,c,d,e,f,g,h,i,j,k,l,m,n,o,p,q,r,s,t,u,v,w,x,y,z"@pt-BR ;
```
Comment (chenejac) [inline home/src/main/resources/rdf/i18n/pt_BR/interface-i18n/firsttime/vitro_UiLabel.ttl:6452]: Should check this with a Portuguese native speaker

**Review by chenejac — CHANGES_REQUESTED** [review]
@litvinovg I don't have any big complain on your code, the implementation is aligned with other Vitro DataGetters implementations. Please find a couple of my tiny comments. 

The PR feature has been tested and it works as it is described with the exception of person's publications.

File: home/src/main/resources/rdf/i18n/de_DE/interface-i18n/firsttime/vitro_UiLabel.ttl (hunk 2 above)
Comment (hauschke) [reply home/src/main/resources/rdf/i18n/de_DE/interface-i18n/firsttime/vitro_UiLabel.ttl:6452]: DIN 5007 should be applied for sorting- https://de.wikipedia.org/wiki/Alphabetische_Sortierung#Deutschland 

Additional letter for Umlauts in the browsing list are not necessary.

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/search/controller/PagedSearchController.java [hunk 4]
```diff
@@ -134,8 +130,12 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
 
     @Override
     protected ResponseValues processRequest(VitroRequest vreq) {
-
-        // There may be other non-html formats in the future
+    	Map<String, List<String>> requestFilters = SearchFiltering.getRequestFilters(vreq);
+        return process(vreq, requestFilters);
+    }
+ 
+	public static ResponseValues process(VitroRequest vreq, Map<String, List<String>> requestFilters) {
```
[the remark this answers is not in the document]
Comment (litvinovg) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/search/controller/PagedSearchController.java]: Fixed.

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/search/controller/PagedSearchController.java [hunk 5]
```diff
@@ -176,18 +176,12 @@ protected ResponseValues processRequest(VitroRequest vreq) {
...
-                filter.setInputText(SearchFiltering.getFilterInputText(vreq, filter.getId()));
-                filter.setRangeValues(SearchFiltering.getFilterRangeText(vreq, filter.getId()));
-            }
-            Map<String, List<String>> requestFilters = SearchFiltering.getRequestFilters(vreq);
-            if (log.isDebugEnabled()) {
-                log.debug(getSpentTime(startTime) + "ms spent after getRequestFilters.");
-            }
-            SearchFiltering.setSelectedFilters(filterConfigurationsByField, requestFilters);
+            
+        	SearchFiltering.setSelectedFilters(filterConfigurationsByField, requestFilters);
```
[the remark this answers is not in the document]
Comment (litvinovg) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/search/controller/PagedSearchController.java]: Fixed

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/search/controller/PagedSearchController.java [hunk 6]
```diff
@@ -313,13 +307,13 @@ protected ResponseValues processRequest(VitroRequest vreq) {
         } catch (Throwable e) {
             return doSearchError(e, format);
         }
-    }
+	}
```
[the remark this answers is not in the document]
Comment (litvinovg) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/search/controller/PagedSearchController.java]: Fixed

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/edit/n3editing/configuration/preprocessors/utils/ProcessDataGetterN3Utils.java [hunk 7]
```diff
@@ -17,26 +17,26 @@ public class ProcessDataGetterN3Utils {
...
     		try {
     			ProcessDataGetterN3 pn = instantiateClass(processorClass, jsonObject);
     			return pn;
     		} catch(Exception ex) {
     			log.error("Exception occurred in trying to get processor class for n3 for " + dataGetterClass, ex);
     			return null;
-    		}
+    		} 
     	}
+    	log.error(ProcessDataGetterN3Map.class.getSimpleName() + " doesn't contain processor class for n3 for " + dataGetterClass);
```
[the remark this answers is not in the document]
Comment (litvinovg) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/edit/n3editing/configuration/preprocessors/utils/ProcessDataGetterN3Utils.java]: Fixed

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/utils/dataGetter/SearchFilterValuesDataGetter.java (hunk 1 above)
Comment (litvinovg) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/utils/dataGetter/SearchFilterValuesDataGetter.java]: Fixed

**Review by chenejac — CHANGES_REQUESTED** [review]
@litvinovg please check my comments. Moreover, the search filter by publication year requests rebuilding search index in the case some record publication date has been changed? I was expected it might be automatically done in the background.

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/search/controller/SearchFilterGroup.java [hunk 8]
```diff
@@ -6,8 +6,7 @@ public class SearchFilterGroup {
 
     private String id;
     private String label;
-    private boolean isPublic = false;
-    private boolean hidden = true;
+    private boolean display = false;
```
[the remark this answers is not in the document]
Comment (litvinovg) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/search/controller/SearchFilterGroup.java]: Fixed

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/search/controller/SearchFilter.java [hunk 9]
```diff
@@ -40,15 +55,16 @@ public class SearchFilter {
...
     private boolean inputRegex = false;
-
     private boolean facetsRequired;
-
+    private boolean descendingOrder;
     private String type = FILTER;
     private String rangeText = "";
     private String rangeInput = "";
-    private boolean hidden = false;
+    private boolean display = false;
```
[the remark this answers is not in the document]
Comment (litvinovg) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/search/controller/SearchFilter.java]: Fixed

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/edit/n3editing/configuration/preprocessors/utils/ProcessSearchFilterValuesDataGetterN3.java [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+    public List<String> retrieveUrisOnForm(int counter) {
+        List<String> urisOnForm = new ArrayList<String>();
+        urisOnForm.add(getVarName("filterUri", counter));
+        urisOnForm.add(getVarName(classTypeVarBase, counter));
+        return urisOnForm;
+    }
+
+    public List<FieldVTwo> retrieveFields(int counter) {
+        List<FieldVTwo> fields = new ArrayList<FieldVTwo>();
+        fields.add(new FieldVTwo().setName(getVarName("filterUri", counter)));
```
[the remark this answers is not in the document]
Comment (litvinovg) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/edit/n3editing/configuration/preprocessors/utils/ProcessSearchFilterValuesDataGetterN3.java]: Fixed

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/search/controller/FilterValue.java [hunk 11]
```diff
@@ -14,18 +14,14 @@ public class FilterValue {
 
     private boolean isDefaultValue;
 
-    private boolean publiclyAvailable = true;
+    private boolean display = false;
```
[the remark this answers is not in the document]
Comment (litvinovg) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/search/controller/FilterValue.java]: Fixed

File: webapp/src/main/webapp/templates/freemarker/body/menupage/browseSearchFilterValues.ftl [hunk 12]
```diff
@@ -0,0 +1,260 @@
...
+                        </#if>
+                    </ul>
+                    <@alphabeticalIndexLinks />
+                </nav>
+                <section id="individuals-in-class" role="region">
+                    <@printPagingLinks />
+                    <div id="browsing-options">
+                        <@showSortOptions />
+                        <@sl.showHits "filter-form" />
+                        <img id="downloadIcon" alt="${i18n().download_results}" title="${i18n().download_results}" />
```
[the remark this answers is not in the document]
Comment (chenejac) [reply webapp/src/main/webapp/templates/freemarker/body/menupage/browseSearchFilterValues.ftl:47]: ok

**Review by chenejac — APPROVED** [review]
@litvinovg well done. Thanks

**Review by chenejac — DISMISSED** [review]

**Comment by litvinovg** [comment]
@brianjlowe I have update test instructions, please try it when you have time.

**Review by chenejac — COMMENTED** [review]
Looks good, I have noticed only one issue which I will state below, but I believe it is an edge case which can be issue only if VIVO is quite empty with data. 

1. On a browse custom page if someone click on a letter (for instance G) and if there is no any facet on the left side, the layout is ruined.

**Comment by brianjlowe** [comment]
Thanks; the feature is now working for me, but I do see some UI issues that I think should be addressed before merging:

1. **Lack of a "there are no results for ..." message when no results are returned.**
Screenshot 2025-05-07 at 13 14 49
When an alpha filter is applied, it should use the same language as the existing classgroup browse:
"There are no entries starting with X.

Please try another letter or browse all."
There should also be a message if there are no results at all. (The main search page says "0 results found", which may be good enough.)

2. **Confusing UI when an item is selected from the left column.**
Screenshot 2025-05-07 at 13 15 41

The black arrow widget that is added to the left of the selected item is similar to the green widget that is used in the classgroup browse but the behavior is quite different. The selected row in the existing classgroup browse is essentially a radio button because the only way to move the widget is to select something different, while on this new page the selected row is semantically a checkbox because it can be clicked again to be deselected and return the search to its previous filter state. If we want to display the selected filter values inline, I think it would be far better to use actual checkboxes to make it clear to the user what will happen if they click the values again. 

It might be worth considering displaying the applied filter values above the list of remaining values and giving them obvious clickable Xs to show that they can be removed.

The secondary disadvantage of the arrow widget icon is that is looks like something that can be clicked to expand a tree. In fact, when I first looked at the page I thought clicking the row would turn the widget to a down-pointing arrow and reveal additional options below, and I was surprised when the row instead went away entirely.

Additionally, I think putting the filter name and colon ("Type:" in this example) is confusing because nothing else on the page says Type. Perhaps it would be useful if there were multiple filters on the page and one were labeled "Type," but then it should only be necessary if all of the applied values for all of the filters are grouped together at the top of the page, not displayed inline as they are now.

3. **Extra space between results, ragged alignment of results per page option, and differing font**
Screenshot 2025-05-07 at 13 15 01

There is a double line and lots of wasted space between each search result. This differs from the existing classgroup browse, where only a single line and modest spacing is used.

It's a minor point, but it would be nice if the "30 Results per page" were aligned with the text in the two rows above it instead of sitting to the left of it.  The number of results dropdown also uses a different font from the links around it.  The R in results looks odd capitalized, but perhaps that's a limitation in the i18n strings.

Overall, I think it would be preferable to have the results per page, sorting and download links look identical on the regular search results pages and these pages instead of making the user look for them in two different locations.

4. **Inconsistent capitalization in page management options**
Screenshot 2025-05-07 at 13 16 23
It looks like "Select a type" (Only first word is capitalized) had already broken from the rest of the pack, but it would be good to align all of them now rather than propagating the inconsistency with the new option. I personally think changing the others to match the "Select a type" and "Browse search filter results" style would be preferable. (And while we're at it, the existing "Sparql" should be corrected to "SPARQL".)

5. Word "undefined" showing in page management.
Screenshot 2025-05-07 at 13 16 32
There may be nothing to show in that heading at this moment, but if so it should just be blank; the "undefined" suggests to the user that something is broken.

**Comment by brianjlowe** [comment]
Looking very good now.  I see just two tiny things: (1) Can some space be added above the "There are no entries starting with" message so that it looks balanced vertically in the frame? Right now it's pretty tight against the top border. And (2): Can we add a period after the first sentence "There are no entries starting with x**.**" so that it is parallel with the sentence below it?
Screenshot 2025-05-14 at 18 21 36

**Review by brianjlowe — APPROVED** [review]

**Comment by litvinovg** [comment]
Force-pushed to resolve merge conflict with recently merged PRs. No changes related to this PR added.

**Review by chenejac — APPROVED** [review]
@litvinovg thanks for this significant contribution.

**Review by brianjlowe — APPROVED** [review]
